### PR TITLE
Support the Attachment object💡 #2616

### LINKED ISSUE #2616: Support the Attachment object💡
Sub-task of #1956.

Implement support of the Attachment object with tests and the common functionality available for the other objects.

### PR DESCRIPTION
**Pull request by robertbasti**
[//]: # (Request Template Witsml Explorer)

[//]: # (Thank you for contributing to WITSML Explorer! Before submitting this PR, please fill in the following template.)

## Fixes

[//]: # (Write the GitHub issue number starting with #, or, for Equinor only, the Jira issue number starting with WE-)

This pull request fixes #2616

## Description

[//]: # (Please include a written summary of the changes and which issue is fixed. List any dependencies that are required for this change._)

added support of the Attachment object with tests and the common functionality 

## Type of change

[//]: # (Mark any of the types of change that apply.)

* [ ] Bugfix
* [ ] New feature (non-breaking change which adds functionality)
* [x] Enhancement of existing functionality
* [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
* [ ] This change requires a documentation update

## Impacted Areas in Application

[//]: # (List general components of the application that this PR will affect)

* [ ] Frontend
* [x] API
* [x] WITSML
* [ ] Desktop
* [ ] Other (please describe)

## Checklist:

[//]: # (Please tick all the boxes or remove the ones that aren't needed and explain why)

*Communication*
* [ ] I have made corresponding changes to the documentation
* [ ] PR affects application security

*Code quality*
* [x] I have self-reviewed my code
* [x] No new warnings are generated

*Test coverage*
* [ ] New code is covered by passing tests

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

File: Tests/WitsmlExplorer.IntegrationTests/Resources/attachment.xml
Comment (eliasbruvik) [inline Tests/WitsmlExplorer.IntegrationTests/Resources/attachment.xml:1]: Have you checked how the image looks? If this is an example from something you found on the server, I think we should be a bit careful before using it.

File: Src/Witsml/Data/WitsmlAttachment.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
+using System.Collections.Generic;
+using System.Xml.Serialization;
+
+using Witsml.Data.Measures;
```
Comment (eliasbruvik) [inline Src/Witsml/Data/WitsmlAttachment.cs:4]: Remove unused libraries (also from the other files)

File: Src/Witsml/Data/WitsmlAttachment.cs
Comment (eliasbruvik) [inline Src/Witsml/Data/WitsmlAttachment.cs:1]: This file is missing _subObjectReference_, _md_, _mdBit_ and _param_. Look for the file _grp_attachment.xsd_ to find the documentation.

File: Src/WitsmlExplorer.Api/Models/Attachment.cs
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Models/Attachment.cs:1]: Also add the missing fields here.

File: Src/WitsmlExplorer.Api/Models/EntityType.cs
Comment (eliasbruvik) [inline Src/WitsmlExplorer.Api/Models/EntityType.cs:1]: Can you also add Attachment to the other mappings in this file?

File: Tests/WitsmlExplorer.Api.Tests/Models/EntityTypeTests.cs
Comment (eliasbruvik) [inline Tests/WitsmlExplorer.Api.Tests/Models/EntityTypeTests.cs:1]: Also add Attachment to the other relevant tests here.

**Review by eliasbruvik — CHANGES_REQUESTED** [review]

File: Src/Witsml/Data/WitsmlAttachment.cs (hunk 1 above)
Comment (robertbasti) [reply Src/Witsml/Data/WitsmlAttachment.cs:4]: done

File: Tests/WitsmlExplorer.IntegrationTests/Resources/attachment.xml
Comment (robertbasti) [reply Tests/WitsmlExplorer.IntegrationTests/Resources/attachment.xml:1]: replaced with some encoded icon file

File: Src/Witsml/Data/WitsmlAttachment.cs
Comment (robertbasti) [reply Src/Witsml/Data/WitsmlAttachment.cs:1]: done

File: Src/WitsmlExplorer.Api/Models/EntityType.cs
Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Models/EntityType.cs:1]: done

File: Tests/WitsmlExplorer.Api.Tests/Models/EntityTypeTests.cs
Comment (robertbasti) [reply Tests/WitsmlExplorer.Api.Tests/Models/EntityTypeTests.cs:1]: done

File: Src/WitsmlExplorer.Api/Models/Attachment.cs
Comment (robertbasti) [reply Src/WitsmlExplorer.Api/Models/Attachment.cs:1]: we will add them to API later when needed

File: Src/Witsml/Data/WitsmlAttachment.cs (hunk 1 above)
Comment (eliasbruvik) [reply Src/Witsml/Data/WitsmlAttachment.cs:4]: Still unused libraries in _Attachment.cs_, _AttachmentQueries.cs_ and _AttachmentService.cs_.

Comment (robertbasti) [reply Src/Witsml/Data/WitsmlAttachment.cs:4]: removed

File: Tests/WitsmlExplorer.IntegrationTests/Resources/attachment.xml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+<attachments xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns="http://www.witsml.org/schemas/1series" xsi:schemaLocation="http://www.witsml.org/schemas/1series  ../xsd_schemas/obj_attachment.xsd" version="1.4.1.1">
```
Comment (eliasbruvik) [inline Tests/WitsmlExplorer.IntegrationTests/Resources/attachment.xml]: Simplify this to ```<attachments version="1.4.1.1" xmlns="http://www.witsml.org/schemas/1series">``` to keep it similar to the other tests.

File: Tests/WitsmlExplorer.IntegrationTests/Resources/attachment.xml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  <attachment uidWell="8c77de13-4fad-4b2e-ba3d-7e6b0e35a394" uidWellbore="ae75db48-5cef-4bd1-9ddf-6035b0d2cd49" uid="attachmentUid">
+    <nameWell>my well</nameWell>
+    <nameWellbore>my wellbore</nameWellbore>
+    <name>1</name>
+    <objectReference object="rig" uidRef="Rig_1">Rig 1</objectReference>
+    <subObjectReference object="3" uidRef="4">subObjRef</subObjectReference>
+    <md uom="m">123</md>
+    <mdBit uom="m">456</mdBit>
+    <typeMessage>type</typeMessage>
```
Comment (eliasbruvik) [inline Tests/WitsmlExplorer.IntegrationTests/Resources/attachment.xml]: Remove typeMessage as it does not belong in an attachment.

File: Tests/WitsmlExplorer.IntegrationTests/Resources/attachment.xml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    <commonData>
+      <dTimCreation>2023-05-17T20:24:07.064Z</dTimCreation>
+      <dTimLastChange>2023-05-17T20:24:07.064Z</dTimLastChange>
```
Comment (eliasbruvik) [inline Tests/WitsmlExplorer.IntegrationTests/Resources/attachment.xml]: Remove ```dTimCreation``` and ```dTimLastChange``` as they are auto-generated by the server and cannot be set in the query.

**Review by eliasbruvik — CHANGES_REQUESTED** [review]

File: Tests/WitsmlExplorer.IntegrationTests/Resources/attachment.xml (hunk 2 above)
Comment (robertbasti) [reply Tests/WitsmlExplorer.IntegrationTests/Resources/attachment.xml]: done

File: Tests/WitsmlExplorer.IntegrationTests/Resources/attachment.xml (hunk 3 above)
Comment (robertbasti) [reply Tests/WitsmlExplorer.IntegrationTests/Resources/attachment.xml]: done

File: Tests/WitsmlExplorer.IntegrationTests/Resources/attachment.xml (hunk 4 above)
Comment (robertbasti) [reply Tests/WitsmlExplorer.IntegrationTests/Resources/attachment.xml]: done

**Review by eliasbruvik — APPROVED** [review]
*Reviewer checklist*
- [x] Read issue and PR
- [x] Pulled branch and manually tested
- [x] Verify that PR resolves issue
- [x] Reviewed the code

**Pull request merged**